File: First Bad Version.java
// https://leetcode.com/problems/first-bad-version/

/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        
        // Create left and right pointer
        int leftPtr = 1;
        int rightPtr = n;
        
        // Move pointers until 1st bad version is found
        while (leftPtr <= rightPtr) {
            
            // Select mid point
            int midPtr = leftPtr + (rightPtr - leftPtr) / 2;
            
            // If midPtr is bad, then move rightPtr to element before mid point
            // This will narrow the range of possible bad versions
            if (isBadVersion(midPtr) == true) {
                rightPtr = midPtr - 1;
                continue;
            }
            
            // Update leftPtr
            leftPtr = midPtr + 1;
        }
      
        // Once while-loop ends, return leftPtr because it's at 1st bad version
        return leftPtr;
    }
}

File: First Bad Version.py
# https://leetcode.com/problems/first-bad-version/

# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        
        # Create left and right pointers
        leftPtr = 1
        rightPtr = n
        
        # Stop while-loop if pointers overlap
        while rightPtr >= leftPtr:
            
            # Get middle version number
            midPtr = leftPtr + (rightPtr - leftPtr) // 2
            
            # Check if middle version is bad
            if isBadVersion(midPtr):
                
                # If true, move rightPtr to one unit before midPtr
                rightPtr = midPtr - 1
                continue
                
            else:
                
                # If false, move leftPtr to one unit after midPtr
                leftPtr = midPtr + 1
                
        # Return 1st bad version
        return leftPtr
